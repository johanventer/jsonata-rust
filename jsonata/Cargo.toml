[package]
name = "jsonata"
version = "0.1.0"
authors = ["Johan Venter <johan.venter@gmail.com>"]
edition = "2021"

[dependencies]
jsonata-errors = { path = "../jsonata-errors" }

chrono = "0.4"
structopt = "0.3"
bitflags = "1.3.2"
bumpalo = { version = "3.9.1", features = ["collections", "boxed"] }
hashbrown = { version = "0.12.0", features = ["bumpalo"] }

[dev-dependencies]
test-case = "1.2.0"
test-generator = "0.3"
