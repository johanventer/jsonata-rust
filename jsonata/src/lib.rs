// TODO: Fix visibility of all these modules, they're all pub for now
pub mod ast;
pub mod evaluator;
pub mod frame;
pub mod functions;
pub mod json;
pub mod parser;
pub mod symbol;
pub mod tokenizer;
pub mod value;

pub use functions::FunctionContext;
pub use jsonata_errors::{Error, Result};
pub use value::Value;

use bumpalo::Bump;

use ast::Ast;
use evaluator::Evaluator;
use frame::Frame;
use functions::*;
use value::ArrayFlags;

pub struct JsonAta<'a> {
    ast: Ast,
    frame: Frame<'a>,
    arena: Bump,
}

impl<'a> JsonAta<'a> {
    pub fn new(expr: &str) -> Result<JsonAta<'a>> {
        Ok(Self {
            ast: parser::parse(expr)?,
            frame: Frame::new(),
            arena: Bump::new(),
        })
    }

    pub fn ast(&self) -> &Ast {
        &self.ast
    }

    pub fn assign_var<'other>(&'other self, name: &str, value: &'other Value<'other>)
    where
        'other: 'a,
    {
        self.frame.bind(name, value)
    }

    pub fn evaluate(&'a self, input: Option<&str>) -> Result<&'a Value<'a>> {
        let input = match input {
            Some(input) => json::parse(input, &self.arena).unwrap(),
            None => Value::undefined(),
        };

        // If the input is an array, wrap it in an array so that it gets treated as a single input
        let input = if input.is_array() {
            Value::wrap_in_array(&self.arena, input, ArrayFlags::WRAPPED)
        } else {
            input
        };

        macro_rules! bind {
            ($name:literal, $new:ident, $fn:ident) => {
                self.frame.bind($name, Value::$new(&self.arena, $name, $fn));
            };
        }

        self.frame.bind("$", input);
        bind!("lookup", nativefn2, fn_lookup);
        bind!("append", nativefn2, fn_append);
        bind!("boolean", nativefn1, fn_boolean);
        bind!("filter", nativefn2, fn_filter);
        bind!("string", nativefn1, fn_string);
        bind!("count", nativefn1, fn_count);
        bind!("not", nativefn1, fn_not);
        bind!("uppercase", nativefn1, fn_uppercase);
        bind!("lowercase", nativefn1, fn_lowercase);
        bind!("substring", nativefn3, fn_substring);
        bind!("abs", nativefn1, fn_abs);
        bind!("max", nativefn1, fn_max);
        bind!("min", nativefn1, fn_min);
        bind!("ceil", nativefn1, fn_ceil);
        bind!("floor", nativefn1, fn_floor);
        bind!("sum", nativefn1, fn_sum);

        let chain_ast = parser::parse("function($f, $g) { function($x){ $g($f($x)) } }")?;
        let evaluator = Evaluator::new(chain_ast, &self.arena);
        evaluator.evaluate(&self.ast, input, &self.frame)
    }
}
