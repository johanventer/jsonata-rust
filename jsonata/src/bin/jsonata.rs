use std::path::PathBuf;
use structopt::StructOpt;

use jsonata::JsonAta;

/// A command line JSON processor using JSONata
#[derive(StructOpt)]
#[structopt(name = "jsonata")]
struct Opt {
    /// Parse the given expression, print the AST and exit
    #[structopt(short, long)]
    ast: bool,

    /// File containing the JSONata expression to evaluate (overrides expr on command line)
    #[structopt(short, long, parse(from_os_str))]
    expr_file: Option<PathBuf>,

    /// Input JSON file (if not specified, STDIN)
    #[structopt(short, long, parse(from_os_str))]
    input_file: Option<PathBuf>,

    /// JSONata expression to evaluate
    expr: Option<String>,

    /// JSON input
    input: Option<String>,
}

fn main() {
    let opt = Opt::from_args();

    let expr = match opt.expr_file {
        Some(expr_file) => {
            let expr = std::fs::read(expr_file).expect("Could not read expression input file");
            String::from_utf8_lossy(&expr).to_string()
        }
        None => opt.expr.expect("No JSONata expression provided"),
    };

    let jsonata = JsonAta::new(&expr);

    match jsonata {
        Ok(jsonata) => {
            if opt.ast {
                println!("{:#?}", jsonata.ast());
                return;
            }

            let input = match opt.input_file {
                Some(input_file) => {
                    std::fs::read_to_string(input_file).expect("Could not read the JSON input file")
                }
                None => opt.input.unwrap_or_else(|| "{}".to_string()),
            };

            match jsonata.evaluate(Some(&input)) {
                Ok(result) => println!("{}", result.pretty(4)),
                Err(error) => println!("{}", error),
            }
        }
        Err(error) => println!("{}", error),
    }
}
