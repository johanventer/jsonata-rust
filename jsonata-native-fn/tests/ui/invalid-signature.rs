use jsonata_native_fn::jsonata;

#[jsonata("yyyy")]
fn test() {}

fn main() {}
