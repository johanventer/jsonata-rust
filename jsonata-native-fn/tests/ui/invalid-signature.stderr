error: F0402 @ Signature does not start with '<'

         = help: Signature parsing failed, check the signature syntax

 --> tests/ui/invalid-signature.rs:3:11
  |
3 | #[jsonata("yyyy")]
  |           ^^^^^^
